File: 38.R
# Pandigital multiples

max_pandigital <- 0
number <- 0
got_max <- FALSE
while (!got_max) {
    number <- number + 1
    if ( grepl('0', number) ) {
        next
    }
    n <- 1
    while ( TRUE ) {
        n <- n + 1
        concat_product <- as.character(number)
        for ( multiple in seq.int(2, n) ) {
            concat_product <- paste0(concat_product, multiple * number,
                                     collapse='')
        }
        if ( nchar(concat_product) < 9 ) {
            next
        }
        if ( nchar(concat_product) > 9 && n == 2 ) {
            got_max <- TRUE
            break
        }
        if ( nchar(concat_product) > 9 ) {
            break
        }
        if ( grepl('0', concat_product) ) {
            next
        }
        digits <- unique(as.integer(strsplit(concat_product, split=NULL)[[1]]))
        if ( length(digits) != 9 ) {
            next
        }
        if ( as.integer(concat_product) > max_pandigital ) {
            max_pandigital <- as.integer(concat_product)
        }
    }
}

cat(max_pandigital, fill=TRUE)

File: 38.js
#!/usr/bin/env node

// Pandigital multiples

max_pandigital = 0;
number = 0;
got_max = false;

while (!got_max) {
    number++;
    if (number.toString().indexOf('0') >= 0) {
        continue;
    }
    var n = 1;
    while (true) {
        n++;
        var concat_product = number.toString();
        for (var multiple = 2; multiple <= n; multiple++) {
            concat_product += (multiple * number).toString();
        }
        if (concat_product.length < 9) {
            continue;
        }
        if (concat_product.length > 9 && n == 2) {
            got_max = true;
            break;
        }
        if (concat_product.length > 9) {
            break;
        }
        if (concat_product.indexOf('0') >= 0) {
            continue;
        }
        digits = unique(concat_product.split(''));
        if (digits.length != 9) {
            continue;
        }
        if (parseInt(concat_product) > max_pandigital) {
            max_pandigital = parseInt(concat_product);
        }
    }
}

console.log(max_pandigital);

function unique(array) {
    return array.reduce(function(prev, cur) {
        if (prev.indexOf(cur) < 0) {
            prev.push(cur);
        }
        return prev;
    }, []);
}
